contracts/token/ERC20WithPermit.sol

```diff
@@ -1,7 +1,232 @@
 // SPDX-License-Identifier: MIT
 
-pragma solidity 0.8.4;
+pragma solidity ^0.8.5;
 
-contract Erc20WithPermit {
-    // TODO will be implemented separately
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./IERC20WithPermit.sol";
+import "./IReceiveApproval.sol";
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+contract ERC20WithPermit is IERC20WithPermit, Ownable {
+    mapping(address => uint256) public override balanceOf;
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    mapping(address => uint256) public override nonces;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    // keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 public constant override PERMIT_TYPEHASH =
+        0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+
+    uint256 public override totalSupply;
+
+    string public override name;
+    string public override symbol;
+
+    uint8 public constant override decimals = 18;
+
+    constructor(string memory _name, string memory _symbol) {
+        name = _name;
+        symbol = _symbol;
+
+        cachedChainId = chainId();
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    function transfer(address recipient, uint256 amount)
+        external
+        override
+        returns (bool)
+    {
+        _transfer(msg.sender, recipient, amount);
+        return true;
+    }
+
+    function transferFrom(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) external override returns (bool) {
+        uint256 currentAllowance = allowance[sender][msg.sender];
+        if (currentAllowance != type(uint256).max) {
```

Comment:
So if `currentAllowance` is equal `type(uint256).max` we won't get into this if statement. We won't decrease the allowance with the `_approve` function, but execute `_transfer` only. Is this some feature that allows unlimited tokens transfers? 
I couldn't find it in [OZ ERC20](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L149), is it a tweak of [UniswapV2ERC20 solution](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2ERC20.sol#L74)?
If this is an intentional feature we should mention it in the docs. 

By the way, we're missing docs for all the functions, so maybe it would be worth adding them?

Reply:
Correct. It is from [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612):
> The deadline argument can be set to uint(-1) to create Permits that effectively never expire.

I don't think OpenZeppelin implements this part. I guess it's because EIP1612 implementation is still treated as a draft in their repo.

Reply:
EIP-2612 is saying about unlimited time for the permit (deadline = uint(-1)), but here we do the unlimited allowed amount.

Reply:
Correct. EIP-2612 says:

> The deadline argument can be set to uint(-1) to create Permits that effectively never expire.

It says nothing about unit(-1) approvals but [Uniswap LP token](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2ERC20.sol#L74) mentioned as a reference implementation of EIP-2612 implements them. Same for [Maker DAO's DAI](https://github.com/makerdao/dss/blob/3f30552a586264b32ebdb5bac94ba67020282e53/src/dai.sol#L80). I think it makes sense given that permit and approval are siblings and if we can set unit(-1) deadline, why not to set unit(-1) approval. EIP-2612 is still in draft, so we'll have some small slight differences between tokens.

Regarding docs, we have them defined on the interface and EIP-20 is well-known so I don't think having them on the token would be a huge improvement.

If we want to add docs, I'd have [them the same way as OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/566a774222707e424896c0c390a84dc3c13bdcb2/contracts/token/ERC20/ERC20.sol#L124-L130) and just list subtle details, like unit(-1) approval amount. Does it work?

Reply:
> Regarding docs, we have them defined on the interface and EIP-20 is well-known so I don't think having them on the token would be a huge improvement.

I agree to the standard `ERC20` functions, but the `ERC20WithPermit` has some significant changes in `transferFrom` which definitely should be mentioned in the documentation.

Reply:
> It says nothing about unit(-1) approvals but Uniswap LP token mentioned as a reference implementation of EIP-2612 implements them. Same for Maker DAO's DAI. I think it makes sense given that permit and approval are siblings and if we can set unit(-1) deadline, why not to set unit(-1) approval. EIP-2612 is still in draft, so we'll have some small slight differences between tokens.

So we will have two unlimited possibilities for `deadline` and `amount`.  We need to be clear in the documentation.
Also keeping mentioning `uint(-1)` is confusing as it's not reflected in the code as it's not supported in Solidity 0.8. We should prioritize mentioning max uint256 in the documentation with a brief reference that it replaces `uint(-1)`.

Reply:
I'm going to be a butt here: please let's add solid documentation, not half-assed docs like most contracts. Having to reference ERC20 to understand mechanics constantly is… Not great. Doesn't need to block any work, but we should look at each contract we formalize in our own code as an opportunity to raise the bar.

Reply:
https://github.com/thesis/solidity-contracts/pull/3/commits/dec702f9f8fb532d4f1d83d8f6aa56e349b8f6b3